crates/sui-core/src/storage.rs

```diff
@@ -0,0 +1,164 @@
+use sui_types::messages_checkpoint::CheckpointDigest;
+use sui_types::messages_checkpoint::CheckpointSequenceNumber;
+use sui_types::messages_checkpoint::VerifiedCheckpoint;
+use sui_types::storage::ReadStore;
+use sui_types::storage::WriteStore;
+use typed_store::Map;
+
+use crate::authority::AuthorityStore;
+use crate::checkpoints::CheckpointService;
+use crate::epoch::committee_store::CommitteeStore;
+
+#[derive(Clone)]
+pub struct RocksDbStore {
+    authority_store: Arc<AuthorityStore>,
+    committee_store: Arc<CommitteeStore>,
+    checkpoint_store: Arc<CheckpointService>,
+}
+
+impl RocksDbStore {
+    pub fn new(
+        authority_store: Arc<AuthorityStore>,
+        committee_store: Arc<CommitteeStore>,
+        checkpoint_store: Arc<CheckpointService>,
+    ) -> Self {
+        Self {
+            authority_store,
+            committee_store,
+            checkpoint_store,
+        }
+    }
+}
+
+impl ReadStore for RocksDbStore {
+    type Error = typed_store::rocks::TypedStoreError;
+
+    fn get_checkpoint_by_digest(
+        &self,
+        digest: &CheckpointDigest,
+    ) -> Result<Option<VerifiedCheckpoint>, Self::Error> {
+        self.checkpoint_store.get_checkpoint_by_digest(digest)
+    }
+
+    fn get_checkpoint_by_sequence_number(
+        &self,
+        sequence_number: CheckpointSequenceNumber,
+    ) -> Result<Option<VerifiedCheckpoint>, Self::Error> {
+        self.checkpoint_store
+            .get_checkpoint_by_sequence_number(sequence_number)
+    }
+
+    fn get_highest_verified_checkpoint(&self) -> Result<Option<VerifiedCheckpoint>, Self::Error> {
+        self.checkpoint_store.get_highest_verified_checkpoint()
+    }
+
+    fn get_highest_synced_checkpoint(&self) -> Result<Option<VerifiedCheckpoint>, Self::Error> {
+        self.checkpoint_store.get_highest_synced_checkpoint()
+    }
+
+    fn get_checkpoint_contents(
+        &self,
+        digest: &CheckpointContentsDigest,
+    ) -> Result<Option<CheckpointContents>, Self::Error> {
+        self.checkpoint_store.get_checkpoint_contents(digest)
+    }
+
+    fn get_committee(&self, epoch: EpochId) -> Result<Option<Committee>, Self::Error> {
+        Ok(self.committee_store.get_committee(&epoch).unwrap())
+    }
+
+    fn get_transaction(
+        &self,
+        digest: &TransactionDigest,
+    ) -> Result<Option<VerifiedCertificate>, Self::Error> {
+        if let Some(transaction) = self
+            .authority_store
+            .epoch_tables()
+            .pending_certificates
+            .get(digest)?
+        {
+            return Ok(Some(transaction.into()));
+        }
+
+        if let Some(transaction) = self
+            .authority_store
+            .perpetual_tables
+            .certificates
+            .get(digest)?
+        {
+            return Ok(Some(transaction.into()));
+        }
+
+        Ok(None)
+    }
+
+    fn get_transaction_effects(
+        &self,
+        digest: &TransactionEffectsDigest,
```

Comment:
previously, we would fetch TransactionEffects by TransactionDigest, and then verify the checksum matches after retrieving it. That seemed to work quite well in practice - are we sure we want to change this pattern now?

Reply:
We should allow for both ways to query effects. If I have an effects digest, then I should be able to use that to get the effects structure.

Reply:
I'm questioning whether that's necessary. In the previous system, that was never necessary, because TransactionEffects are useless on their own. You always have a TransactionDigest because that's what you're trying to do - execute a tx. So I never once ran into a case where I wanted to look up TransactionEffects based on a TransactionEffectsDigest.

YMMV, and I don't feel strongly about this, just trying to point out what my experience was.